channelserver/src/meta.rs

```diff
@@ -142,6 +148,71 @@ fn get_ua(headers: &http::HeaderMap, log: Option<&Addr<logging::MozLogger>>) ->
     None
 }
 
+fn is_trusted_proxy(proxy_list: &[IpNet], host: &str) -> Result<bool, HandlerError> {
+    // Return if an address is NOT part of the allow list
+    let test_addr: IpAddr = match host.parse() {
+        Ok(addr) => addr,
+        Err(e) => return Err(HandlerErrorKind::BadRemoteAddrError(format!("{:?}", e)).into()),
+    };
+    for proxy_range in proxy_list {
+        if proxy_range.contains(&test_addr) {
+            return Ok(true);
+        }
+    }
+    Ok(false)
+}
+
+fn get_remote(
+    peer: &Option<SocketAddr>,
+    headers: &http::HeaderMap,
+    proxy_list: &[IpNet],
+) -> Result<String, HandlerError> {
+    // Actix determines the connection_info.remote() from the first entry in the
+    // Forwarded then X-Fowarded-For, Forwarded-For, then peer name. The problem is that any
+    // of those could be multiple entries or may point to a known proxy, or be injected by the
+    // user. We strictly only check the one header we know the proxy will be sending, working
+    // our way back up the proxy chain until we find the first unexpected address.
+    // This may be an intermediary proxy, or it may be the original requesting system.
+    //
+    if peer.is_none() {
+        return Err(HandlerErrorKind::BadRemoteAddrError("Peer is unspecified".to_owned()).into());
+    }
+    let peer_ip = peer.unwrap().ip().to_string();
+    // if the peer is not a known proxy, ignore the X-Forwarded-For headers
+    if !is_trusted_proxy(proxy_list, &peer_ip)? {
+        return Ok(peer_ip);
+    }
+
+    // The peer is a known proxy, so take rightmost X-Forwarded-For that is not a trusted proxy.
+    match headers.get(HeaderName::from_lowercase("x-forwarded-for".as_bytes()).unwrap()) {
+        Some(header) => {
+            match header.to_str() {
+                Ok(hstr) => {
+                    // Just like `Forward` successive proxies are appeneded to this header.
+                    let mut host_list: Vec<&str> = hstr.split(',').collect();
+                    host_list.reverse();
+                    for host_str in host_list {
+                        let host = host_str.trim().to_owned();
+                        if !is_trusted_proxy(proxy_list, &host)? {
+                            return Ok(host.to_owned());
+                        }
+                    }
+                    Err(HandlerErrorKind::BadRemoteAddrError(format!(
+                        "Could not find remote IP in X-Forwarded-For"
+                    )).into())
+                }
+                Err(err) => Err(HandlerErrorKind::BadRemoteAddrError(format!(
+                    "Unknown address in X-Forwarded-For: {:?}",
```

Comment:
nit: the error message "unknown addresss" doesn't seem quite right here; IIUC this error would be caused by a failure to interpret the X-F-F header as utf8?

Reply:
Well, it could be for any number of reasons, really. Bad characters, invalid address specification, or anything that would make the `.parse()` fail. It's pretty much a catch-all. I suppose "Unparsable entry" might also work.